### DISCUSSION TITLE
Moving to true Continuous Deployment

### DISCUSSION THREAD

**Discussion by MicahGale** [idaholab/MontePy#371] [Ideas]
After #336 is released, deploys should be a lot easier due to the entire process being almost entirely automated, including version numbering.

This makes me think: is there much value in doing deployment staging? I think now once a feature/fix is merged into Develop it could be immediately deployed. This then would mean deleting develop and merging everything into main, and making much more frequent deployments.

Pros
----
* less developer burden
* Bug fixes can ship faster

Cons
----
* Lots of deployments in PyPI/ github
* One less check before final deployment. (I don't think this is valid because how closely looking are people looking at release PRs)

Alpha Testing
-------------

This would renew the need for alpha-testing.
Do we continue with this just being a git based branch that testers clone? Or do we do alpha deployments to GitHub or (inclusive) PyPI?

Thoughts @tjlaboss?

**Comment by MicahGale** [comment]
So I was looking into: can you do pre-release on PyPI: and yes; you can.

See:

* [numpy](https://pypi.org/project/numpy/#history)
* [Stack overflow on pip and pre-release](https://stackoverflow.com/questions/64319173/pre-release-versions-are-not-matched-by-pip-when-using-the-pre-option)
* ~[PEP 440 on pre-releases](https://peps.python.org/pep-0440/#pre-releases)~ [Packaging doc on pre-release](https://packaging.python.org/en/latest/specifications/version-specifiers/#pre-releases)
* [How our version bumping GHA handles pre-releases](https://github.com/mathieudutour/github-tag-action?tab=readme-ov-file#filter-branches)

**Comment by MicahGale** [comment]
One point that was brought up is: will the change log be hard to work with so many releases? If you have to go through github releases I think that will be true. However, if we implement an effective changelog, this should be well mitigated.

**Comment by tjlaboss** [comment]
First impression is that continuous deployment to pre-release versions would be quite nice.

For the changelog, could one do something like:

____

# Major Version N+1

## Minor Release m+1

### pre-release p+1

 * Bug fix d
 * Feature c

#### pre-release p

 * Bug fix b
 * Feature a

# Major Release N

## . . .

**Reply by MicahGale** [reply]
So I was thinking of making a CD release to Alpha would be a huge part of this. In #380 I introduced a changelog. I was thinking of doing some sort of keyword system like you proposed there. Then in CD:

1. get next version from GHA dry-run
2. `sed` the changelog with the proper version number, and commit.
3. Do the actual version change with git tags
4. continue CD
